File: functions/src/models/swapRequest.ts
import { Timestamp } from 'firebase-admin/firestore';

/**
 * A swap request between two flat members.
 * Stored at flats/{flatId}/swapRequests/{requestId}.
 */
export enum SwapRequestStatus {
  Pending = 'pending',
  Accepted = 'accepted',
  Declined = 'declined',
}

export interface SwapRequest {
  id: string;
  requester_uid: string;
  target_task_id: string;
  requester_task_id: string;
  status: SwapRequestStatus;
  created_at: Timestamp;
}

export type SwapRequestData = Omit<SwapRequest, 'id'>;

export function swapRequestFromFirestore(
  id: string,
  data: FirebaseFirestore.DocumentData,
): SwapRequest {
  return {
    id,
    requester_uid: data['requester_uid'] ?? '',
    target_task_id: data['target_task_id'] ?? '',
    requester_task_id: data['requester_task_id'] ?? '',
    status: (data['status'] as SwapRequestStatus) ?? SwapRequestStatus.Pending,
    created_at: data['created_at'] as Timestamp,
  };
}

File: functions_python/models/swap_request.py
"""SwapRequest model: a pending task-swap between two members."""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum
from typing import Any


class SwapRequestStatus(StrEnum):
    """Lifecycle state of a task-swap request."""

    Pending = "pending"
    Accepted = "accepted"
    Declined = "declined"


@dataclass
class SwapRequest:
    """A request from one member to swap tasks with another."""

    id: str
    requester_uid: str
    requester_task_id: str
    target_task_id: str
    status: SwapRequestStatus
    created_at: Any  # Firestore Timestamp


def swap_request_from_firestore(doc_id: str, data: dict[str, Any]) -> SwapRequest:
    """Convert a Firestore swapRequests document dict to a typed SwapRequest."""
    return SwapRequest(
        id=doc_id,
        requester_uid=data.get("requester_uid", ""),
        requester_task_id=data.get("requester_task_id", ""),
        target_task_id=data.get("target_task_id", ""),
        status=SwapRequestStatus(data.get("status", SwapRequestStatus.Pending)),
        created_at=data.get("created_at"),
    )
